File: theme12/bin/preppage.pl
#!/usr/bin/perl

sub file_to_text {
	$fname=$_[0];
	my $txt;
	$txt='';
	if ( !open INPUTDATA, $fname ) {
		die "cannot open $fname: $!";
	}
	while (<INPUTDATA>) {
		$txt = $txt . $_;
	}
	return $txt;
}

$num_args=$#ARGV +1;
if ($num_args != 4) {
	die "\n  Usage: xxx file\n\n";
}

#post hyper code:
$phc="JDHALOWPFHSNC";

#date hyper code:
$dhc="ASDLKJFHG";

#sidebar hyper code:
$shc="KJDSADPFP";

#footbar hyper code:
$fhc="KDLPQHFZVD";

$template_file=$ARGV[0];
$side_file=$ARGV[1];
$foot_file=$ARGV[2];
$post_file=$ARGV[3];

$txtsuffix = '.txt';
$post_filetmp = `basename -s .w $post_file`;
chomp $post_filetmp;
$post_filet =  $post_filetmp . $txtsuffix;

$TEMPTXT=&file_to_text($template_file);
$POSTTXT=&file_to_text($post_file);
$SIDETXT=&file_to_text($side_file);
$FOOTBARTXT=&file_to_text($foot_file);

#creating date:
# fix for unix using mydate.py
#$lstupddaynum=`date +%d -r $post_filet`;
#chomp $lstupddaynum;
#$lstupddayw=`date +%a -r $post_filet`;
#chomp $lstupddayw;
#$lstupdyear=`date +%Y -r $post_filet`;
#chomp $lstupdyear;
#$lstupdmon=`date +%b -r $post_filet`;
#chomp $lstupdmon;

#$date="$lstupddayw $lstupdmon $lstupddaynum, $lstupdyear";

$date = `./bin/mydate.py $post_filet`;
chomp $date;

$title='';
if ($POSTTXT =~ /(title\s*:=\s*{\s*)([^}]*?)(\s*})/s) {
   $title=$2;
	$POSTTXT =~ s/title\s*:=\s*{[^}]*}//sg;
} else {
	if ($POSTTXT =~ /(<h1>\s*)([^:]*)(:*)(\s*<\/h1>)/s ) {
		$title = $2;
	}
}

$author='';
if ($POSTTXT =~ /(author\s*:=\s*{\s*)([^}]*?)(\s*})/s) {
	$author=$2;
	$POSTTXT =~ s/author\s*:=\s*{[^}]*}//sg;
}


$keywords='';
if ($POSTTXT =~ /(keywords\s*:=\s*{\s*)([^}]*?)(\s*})/s) {
	$keywords=$2;
	$POSTTXT =~ s/keywords\s*:=\s*{[^}]*}//sg;
}

$description='';
if ($POSTTXT =~ /(description\s*:=\s*{\s*)([^}]*?)(\s*})/s) {
	$description=$2;
	$POSTTXT =~ s/description\s*:=\s*{[^}]*}//sg;
}

if ($title ne ''){
  $TEMPTXT=~ s/<title>.*<\/title>/<title>$title<\/title>/sg;
}

$tl='"';

if ($keywords ne '') {
 $TEMPTXT =~ s/(<meta name=${tl}keywords${tl} content=$tl)([^$tl]*)($tl\s*\/>)/$1$keywords$3/sg;
}

if ($author ne '') {
 $TEMPTXT =~ s/(<meta name=${tl}author${tl} content=$tl)([^$tl]*)($tl\s*\/>)/$1$author$3/sg;
}


if ($description ne '') {
 $TEMPTXT =~ s/(<meta name=${tl}description${tl} content=$tl)([^$tl]*)($tl\s*\/>)/$1$description$3/sg;
}

$dl_syn='\[dl\]';
$dl_op='<img src="img/dl.svg" />';

$ext_lnk_syn='\[>\]';
$ext_lnk_op='<img style="margin-left:1px;" src="img/external-link.svg" alt="" align="bottom" />';

$toc_syn='\[TOC\]';
$toc_op='<div class="table_of_contents"></div>';

# Creating references:

$num_ref=0;
$ref_text='';
$is_there_ref=0;
$reftbl{"HDQIFZMXCNAPQEDJALFX"}=0;
while ($POSTTXT =~ /(%%\s*)(\w+.*?)(\s*%%)/ ) {
  $is_there_ref=1;
  $lnk_nm =$2;
  if ( ! exists $ref_nums{$lnk_nm} ) {
	$num_ref++;
	$ref_nums{$lnk_nm}=$num_ref;
	$POSTTXT =~ s/(%%\s*)$lnk_nm(\s*%%)/<sup><a href="#ref$lnk_nm" id="cnt$lnk_nm">\[$num_ref\]<\/a><\/sup>/;
	if ($POSTTXT =~ /(\{\s*$lnk_nm\s*:\s*)([^\}]*)(\})/s ) {
		$each_ref = '<a href="#cnt'.$lnk_nm.'" id="ref'.$lnk_nm.'">[' . $num_ref . ']</a>: '. $2;
		$ref_txt = $ref_txt . '<p>' .  $each_ref . '</p>' . "\n";
	   $POSTTXT =~ s/(\{\s*$lnk_nm\s*:\s*)([^\}]*)(\})//sg; 
	} 
  } else{
	  $num = $ref_nums{$lnk_nm};
	  $POSTTXT =~ s/(%%\s*)$lnk_nm(\s*%%)/<sup><a href="#ref$lnk_nm" id="cnt$lnk_nm">\[$num\]<\/a><\/sup>/;
  }
}

if ($is_there_ref==1) {
$POSTTXT = $POSTTXT . '<hr />' . '<h2>References</h2>' . $ref_txt;
}


$TEMPTXT=~ s/$dhc/$date/sg;
$TEMPTXT=~ s/$phc/$POSTTXT/sg;
$TEMPTXT=~ s/$shc/$SIDETXT/sg;
$TEMPTXT=~ s/$fhc/$FOOTBARTXT/sg;
$TEMPTXT=~ s/$ext_lnk_syn/$ext_lnk_op/sg;
$TEMPTXT=~ s/$dl_syn/$dl_op/sg;
$TEMPTXT=~ s/$toc_syn/$toc_op/sg;

$img_op1='<div style="max-width:350px; width:auto; padding:2px; height:auto; text-align:justify; border: solid #BBBBBB 1px;float:right; margin-left:8px;"><img style="display:block; height:auto; width:auto; max-width:320px; margin-left:auto; margin-right:auto;" src="';
$img_op2='" alt="';
$img_op3='" border="0" /><hr />';
$img_op3noline='" border="0" />';
$img_op4='</div>';
$TEMPTXT=~ s/({{}}\(\()(.*?)(\)\))/$img_op1$2$img_op2$img_op3noline$img_op4/sg;
$TEMPTXT=~ s/({{)(.*?)(}}\(\()(.*?)(\)\))/$img_op1$4$img_op2$img_op3$2$img_op4/sg;


$imgc_op1='<div style="margin-left:auto; margin-right:auto; padding:3px; text-align:justify; border:solid #BBBBBB 1px; height:auto; width: auto; max-width:500px;"><img style="display:block; height:auto; width:auto; max-width:470px; margin-left:auto; margin-right:auto;" src="';
$imgc_op2='" alt="';
$imgc_op3='" border="0" /><hr />';
$imgc_op3noline='" border="0" />';
$imgc_op4='</div>';
$TEMPTXT=~ s/({{}}\[\[)(.*?)(\]\])/$imgc_op1$2$imgc_op2$imgc_op3noline$imgc_op4/sg;
$TEMPTXT=~ s/({{)(.*?)(}}\[\[)(.*?)(\]\])/$imgc_op1$4$imgc_op2$imgc_op3$2$imgc_op4/sg;

$r_op1='<div style="max-width:350px; width:auto; padding:2px; height:auto; text-align:justify; border: solid #BBBBBB 1px;float:right; margin-left:8px;">';
$r_op2='</div>';

$TEMPTXT =~ s/(<R>)(.*?)(<\/R>)/$r_op1$2$r_op2/sg;

print $TEMPTXT;

File: theme1/bin/preppage.py
#!/usr/bin/env python3
import dataclasses
import os
import re
from datetime import datetime
import sys
from typing import Dict, Iterable, Tuple


@dataclasses.dataclass(frozen=True)
class Replacements:
    post_token: str = "JDHALOWPFHSNC"
    date_token: str = "ASDLKJFHG"
    sidebar_token: str = "KJDSADPFP"
    footer_token: str = "KDLPQHFZVD"
    download_token: str = "[dl]"
    external_token: str = "[>]"
    toc_token: str = "[TOC]"


META_FIELDS = ("title", "author", "keywords", "description")


def read_text(path: str) -> str:
    with open(path, "r") as handle:
        return handle.read()


def post_source_name(post_path: str) -> str:
    base_name = os.path.basename(post_path)
    if base_name.endswith(".w"):
        base_name = base_name[: -len(".w")]
    return f"{base_name}.txt"


def render_date(post_path: str) -> str:
    try:
        modified_time = os.path.getmtime(post_path)
        return datetime.fromtimestamp(modified_time).strftime("%a %b %d %Y")
    except OSError:
        return ""


def extract_metadata(post_text: str) -> Tuple[Dict[str, str], str]:
    metadata: Dict[str, str] = {}
    cleaned = post_text

    for field in META_FIELDS:
        pattern = rf"{field}\s*:=\s*{{\s*([^}}]*?)\s*}}"
        match = re.search(pattern, cleaned, flags=re.DOTALL)
        if match:
            metadata[field] = match.group(1)
            cleaned = re.sub(pattern, "", cleaned, flags=re.DOTALL)

    if "title" not in metadata:
        h1_match = re.search(r"(<h1>\s*)([^:]*)(:*)(\s*</h1>)", cleaned, flags=re.DOTALL)
        if h1_match:
            metadata["title"] = h1_match.group(2)

    return metadata, cleaned


def update_meta_tags(template: str, metadata: Dict[str, str]) -> str:
    if "title" in metadata:
        title_text = metadata["title"]
        template = re.sub(
            r"<title>.*</title>",
            lambda _match: f"<title>{title_text}</title>",
            template,
            flags=re.DOTALL,
        )

    meta_patterns = {
        "keywords": r'(<meta name="keywords" content=")([^"]*)(" />)',
        "author": r'(<meta name="author" content=")([^"]*)(" />)',
        "description": r'(<meta name="description" content=")([^"]*)(" />)',
    }

    for key, pattern in meta_patterns.items():
        if key in metadata:
            value = metadata[key]
            template = re.sub(
                pattern,
                lambda match, value=value: f"{match.group(1)}{value}{match.group(3)}",
                template,
                flags=re.DOTALL,
            )

    return template


def build_reference_block(post_text: str) -> str:
    ref_numbers: Dict[str, int] = {}
    next_ref = 0
    ref_html = []

    def replace_reference(match: re.Match[str]) -> str:
        nonlocal next_ref
        label = match.group(2)
        if label not in ref_numbers:
            next_ref += 1
            ref_numbers[label] = next_ref
            definition = re.search(rf"(\{{\s*{re.escape(label)}\s*:\s*)([^}}]*)(\}})", post_text, flags=re.DOTALL)
            if definition:
                entry = (
                    f'<a href="#cnt{label}" id="ref{label}">[{next_ref}]</a>: '
                    f"{definition.group(2)}"
                )
                ref_html.append(f"<p>{entry}</p>")
        ref_id = ref_numbers[label]
        return f'<sup><a href="#ref{label}" id="cnt{label}">[{ref_id}]</a></sup>'

    ref_pattern = re.compile(r"(%%\s*)(\w+.*?)(\s*%%)")
    while ref_pattern.search(post_text):
        post_text = ref_pattern.sub(replace_reference, post_text, count=1)
        for label in list(ref_numbers):
            post_text = re.sub(
                rf"(\{{\s*{re.escape(label)}\s*:\s*)([^}}]*)(\}})",
                "",
                post_text,
                flags=re.DOTALL,
            )

    if ref_html:
        return f"{post_text}<hr /><h2>References</h2>{''.join(ref_html)}"

    return post_text


def apply_shortcodes(text: str, tokens: Replacements) -> str:
    replacements = {
        tokens.external_token: (
            '<img style="margin-left:1px;" src="img/external-link.svg" alt="" align="bottom" />'
        ),
        tokens.download_token: '<img src="img/dl.svg" />',
        tokens.toc_token: '<div class="table_of_contents"></div>',
    }
    for source, target in replacements.items():
        text = re.sub(re.escape(source), target, text)
    return text


def apply_image_macros(text: str) -> str:
    right_box = (
        '<div style="max-width:350px; width:auto; padding:2px; height:auto; text-align:justify; '
        'border: solid #BBBBBB 1px;float:right; margin-left:8px;">'
        '<img style="display:block; height:auto; width:auto; max-width:320px; '
        'margin-left:auto; margin-right:auto;" src="'
    )
    right_alt = '" alt="'
    right_caption = '" border="0" /><hr />'
    right_no_caption = '" border="0" />'
    right_close = "</div>"

    text = re.sub(
        r"({{}}\(\()(.*?)(\)\))",
        rf"{right_box}\2{right_alt}{right_no_caption}{right_close}",
        text,
        flags=re.DOTALL,
    )
    text = re.sub(
        r"({{)(.*?)(}}\(\()(.*?)(\)\))",
        rf"{right_box}\4{right_alt}{right_caption}\2{right_close}",
        text,
        flags=re.DOTALL,
    )

    center_box = (
        '<div style="margin-left:auto; margin-right:auto; padding:3px; text-align:justify; '
        'border:solid #BBBBBB 1px; height:auto; width: auto; max-width:500px;">'
        '<img style="display:block; height:auto; width:auto; max-width:470px; '
        'margin-left:auto; margin-right:auto;" src="'
    )
    center_alt = '" alt="'
    center_caption = '" border="0" /><hr />'
    center_no_caption = '" border="0" />'
    center_close = "</div>"

    text = re.sub(
        r"({{}}\[\[)(.*?)(\]\])",
        rf"{center_box}\2{center_alt}{center_no_caption}{center_close}",
        text,
        flags=re.DOTALL,
    )
    text = re.sub(
        r"({{)(.*?)(}}\[\[)(.*?)(\]\])",
        rf"{center_box}\4{center_alt}{center_caption}\2{center_close}",
        text,
        flags=re.DOTALL,
    )

    right_wrap = (
        '<div style="max-width:350px; width:auto; padding:2px; height:auto; text-align:justify; '
        'border: solid #BBBBBB 1px;float:right; margin-left:8px;">'
    )
    text = re.sub(r"(<R>)(.*?)(</R>)", rf"{right_wrap}\2</div>", text, flags=re.DOTALL)

    return text


def replace_placeholders(
    template: str,
    post_text: str,
    sidebar_text: str,
    footer_text: str,
    date_text: str,
    tokens: Replacements,
) -> str:
    replacements = {
        tokens.date_token: date_text,
        tokens.post_token: post_text,
        tokens.sidebar_token: sidebar_text,
        tokens.footer_token: footer_text,
    }
    for source, target in replacements.items():
        template = template.replace(source, target)
    return template


def main() -> None:
    if len(sys.argv) != 5:
        raise SystemExit("\n  Usage: xxx file\n\n")

    tokens = Replacements()
    template_file, side_file, foot_file, post_file = sys.argv[1:5]

    post_source = post_source_name(post_file)
    template = read_text(template_file)
    post_text = read_text(post_file)
    sidebar_text = read_text(side_file)
    footer_text = read_text(foot_file)

    metadata, post_text = extract_metadata(post_text)
    post_text = build_reference_block(post_text)

    date_text = render_date(post_source)
    template = update_meta_tags(template, metadata)

    output = replace_placeholders(
        template,
        post_text,
        sidebar_text,
        footer_text,
        date_text,
        tokens,
    )
    output = apply_shortcodes(output, tokens)
    output = apply_image_macros(output)

    sys.stdout.write(output)


if __name__ == "__main__":
    main()
